proxyserver/middleware/versioned_writes.go

```diff
@@ -0,0 +1,549 @@
+		}
+		if request.Header.Get("X-If-Delete-At") != "" {
+			request.Header.Del("X-If-Delete-At")
+		}
+		break
+	}
+	v.next.ServeHTTP(writer, request)
+	return
+}
+
+func (v *versionedWrites) handleObjectPut(writer http.ResponseWriter, request *http.Request, account, container, versionsContainer, object string) {
+	if request.Header.Get("X-Object-Manifest") != "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	ok, status := v.copyCurrent(writer, request, account, container, versionsContainer, object)
+	if !ok && returnIfStatusError(writer, status) {
+		return
+	}
+	v.next.ServeHTTP(writer, request)
+}
+
+func (v *versionedWrites) handleObject(writer http.ResponseWriter, request *http.Request) {
+	_, account, container, object := getPathParts(request)
+	ctx := GetProxyContext(request)
+	ci, _ := ctx.C.GetContainerInfo(account, container)
+	if ci == nil {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	versionsContainer := ci.SysMetadata["Versions-Location"]
+	versionsMode := ci.SysMetadata["Versions-Mode"]
+	if versionsMode == "" {
+		versionsMode = "stack"
+	}
+	if versionsContainer == "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	if unescapedVersionsContainer, err := url.QueryUnescape(versionsContainer); err == nil {
+		versionsContainer = strings.Split(unescapedVersionsContainer, "/")[0]
+	} else {
+		ctx.Logger.Info(fmt.Sprintf("Bad versions container: %s", versionsContainer))
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	if request.Method == "PUT" {
+		v.handleObjectPut(writer, request, account, container, versionsContainer, object)
+	} else if versionsMode == "history" {
+		v.handleObjectDeleteHistory(writer, request, account, container, versionsContainer, object)
+	} else {
+		v.handleObjectDeleteStack(writer, request, account, container, versionsContainer, object)
+	}
+}
+
+func returnIfStatusError(writer http.ResponseWriter, status int) bool {
+	if status/100 == 2 {
+		return false
+	}
+	if status == 403 {
+		srv.StandardResponse(writer, 403)
+	} else if status/100 == 4 {
+		srv.StandardResponse(writer, 412)
+	} else {
+		srv.StandardResponse(writer, 500)
+	}
+	return true
+}
+
+func (v *versionedWrites) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	apiReq, account, container, object := getPathParts(request)
+	if container == "" && object == "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	if !apiReq || account == "" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+	if GetProxyContext(request).Source == "VW" {
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	if !v.enabled {
+		// Can't set these headers if disabled
+		if (request.Method == "POST" || request.Method == "PUT") &&
+			(request.Header.Get(CLIENT_VERSIONS_LOC) != "" ||
+				request.Header.Get(CLIENT_HISTORY_LOC) != "") {
+
+			srv.SimpleErrorResponse(writer, 412, "Versioned Writes is disabled")
+			return
+		}
+		v.next.ServeHTTP(writer, request)
+		return
+	}
+
+	if object == "" {
```

Comment:
can we check if it is a GET / HEAD at the beginning of function and just .next / return quickly?